File: 卡码/代码随想录/1/11-图论/Go/bfs/LeetCode_0127.go
// 127. 单词接龙
package main

import (
	"github.com/emirpasic/gods/queues/arrayqueue"
	"github.com/emirpasic/gods/sets/hashset"
)

func ladderLength(beginWord string, endWord string, wordList []string) int {
	visited := hashset.New()
	workSet := hashset.New()
	for _, word := range wordList {
		workSet.Add(word)
	}
	if !workSet.Contains(endWord) {
		return 0
	}
	q := arrayqueue.New()
	q.Enqueue(beginWord)
	path := 1
	var changeLetter func(word string) bool
	changeLetter = func(word string) bool {
		for i, _ := range word {
			for j := 'a'; j <= 'z'; j++ {
				newWord := make([]rune, len(word))
				copy(newWord, []rune(word))
				newWord[i] = j
				if string(newWord) == endWord {
					return true
				}
				if workSet.Contains(string(newWord)) && !visited.Contains(string(newWord)) {
					visited.Add(string(newWord))
					q.Enqueue(string(newWord))
				}
			}
		}
		return false
	}
	for q.Size() > 0 {
		size := q.Size()
		for i := 0; i < size; i++ {
			word, _ := q.Dequeue()
			if changeLetter(word.(string)) {
				return path + 1
			}
		}
		path++
	}
	return 0
}

File: 卡码/代码随想录/1/11-图论/Python/bfs/LeetCode_0127.py
#127. 单词接龙
from typing import *
from collections import deque

class Solution:
  def __init__(self):
    self.q:deque[str]=deque()
    self.visited:set[str]=set()
    self.wordSet:set[str]=set()
  def changeLetter(self,word:str,endWord:str)->bool:
    for i in range(len(word)):
      for ch in range(ord('a'),ord('z')+1):
        newWord:str=word[:i]+chr(ch)+word[i+1:]
        if newWord==endWord:
          return True
        if newWord in self.wordSet and newWord not in self.visited:
          self.visited.add(newWord)
          self.q.append(newWord)
    return False
  def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
    self.wordSet=set(wordList)
    if endWord not in self.wordSet:
      return 0
    self.q.append(beginWord)
    path:int=1
    while len(self.q)>0:
      size:int=len(self.q)
      for i in range(size):
        word:str=self.q.popleft()
        if self.changeLetter(word,endWord):
          return path+1
      path+=1
    return 0
      